pkg/util/grpcutil.go

```diff
@@ -0,0 +1,221 @@
+	return &tls.Config{
+		GetConfigForClient: func(info *tls.ClientHelloInfo) (*tls.Config, error) {
+			if info == nil {
+				return nil, errors.New("nil client info passed")
+			}
+
+			config := &tls.Config{
+				MinVersion:    tls.VersionTLS13,
+				Renegotiation: tls.RenegotiateNever,
+				Certificates:  []tls.Certificate{*peerCert},
+				ClientCAs:     certPool,
+				VerifyPeerCertificate: func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+					// Common name check when accepting a connection from a client.
+					if peerName == "" {
+						// All names allowed.
+						return nil
+					}
+
+					if len(verifiedChains) == 0 ||
+						len(verifiedChains[0]) == 0 {
+						return errors.New("no valid certificate")
+					}
+
+					for _, name := range verifiedChains[0][0].DNSNames {
+						if name == peerName {
+							return nil
+						}
+					}
+					// For compatibility - using CN as hostName
+					commonName := verifiedChains[0][0].Subject.CommonName
+					if commonName == peerName {
+						return nil
+					}
+					return fmt.Errorf("certificate is not signed for %q hostname", peerName)
+				},
+			}
+			if peerName != "" {
+				config.ClientAuth = tls.RequireAndVerifyClientCert
+			}
+			return config, nil
+		},
+	}
+}
+
+// ClientTLS prepares the TLS configuration that can be used by a client while connecting to a server
+// with given encoded certificate and private key.
+// peerName must be provided when expecting the server to offer a certificate with that CommonName.
+func ClientTLS(caCert, cert, key []byte, peerName string) (*tls.Config, error) {
+	certPool := x509.NewCertPool()
+	if ok := certPool.AppendCertsFromPEM(caCert); !ok {
+		return nil, fmt.Errorf("failed to append CA certificate to pool")
+	}
+
+	certificate, err := tls.X509KeyPair(cert, key)
+	if err != nil {
+		return nil, err
+	}
+
+	return clientConfig(certPool, &certificate, peerName), nil
+}
+
+// LoadClientTLS prepares the TLS configuration that can be used by a client while connecting to a server.
+// peerName must be provided when expecting the server to offer a certificate with that CommonName. caFile, certFile, and keyFile are all optional.
+func LoadClientTLS(caFile, certFile, keyFile, peerName string) (*tls.Config, error) {
+	certPool, peerCert, err := loadCertificate(caFile, certFile, keyFile)
+	if err != nil {
+		return nil, err
+	}
+
+	return clientConfig(certPool, peerCert, peerName), nil
+}
+
+func clientConfig(certPool *x509.CertPool, peerCert *tls.Certificate, peerName string) *tls.Config {
+	tlsConfig := &tls.Config{
+		MinVersion:    tls.VersionTLS13,
+		Renegotiation: tls.RenegotiateNever,
+		ServerName:    peerName,
+		RootCAs:       certPool,
+	}
+	if peerCert != nil {
+		tlsConfig.Certificates = append(tlsConfig.Certificates, *peerCert)
+	}
+	return tlsConfig
+}
+
+func loadCertificate(caFile, certFile, keyFile string) (certPool *x509.CertPool, peerCert *tls.Certificate, err error) {
+	if certFile != "" || keyFile != "" {
```

Comment:
This should be `if certFile != "" && keyFile != ""`? otherwise it's unable to create the server/client key pair.

Reply:
You are correct to create a cert/key pair one needs to provide both files:
- providing both files -> success
- providing only 1 of the files will lead to an error return --> error
- Providing 0 of the files is a valid use case though, so that's why it's || instead of and. --> success

Reply:
ref: https://github.com/longhorn/longhorn-instance-manager/pull/109/files#r864656548